GSoC 2020: add style options for multimeasure rests

IsaacWeiss opened the pull request:
Resolves: https://musescore.org/en/user/101731/blog/2020/06/01/gsoc-2020-day-1-it-begins and https://musescore.org/en/node/44846

Old-style multimeasure rests: 
![old-style](https://user-images.githubusercontent.com/11616153/84464582-ec453900-ac42-11ea-959d-5251a340237c.png)

New Style page with new options:
![Screen Shot 2020-06-14 at 4 08 30 PM](https://user-images.githubusercontent.com/11616153/84603035-63ecb100-ae59-11ea-90e5-59d1b7972a5b.png)

Font-specific settings automatically applied:
![Bravura font style](https://user-images.githubusercontent.com/11616153/83712317-7c580280-a5f3-11ea-8877-4ab5b4b8097d.gif)

- [X] I signed [CLA](https://musescore.org/en/cla)
- [X] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/Documentation/blob/master/WorkflowAndGuidelines/CodeGuidelines.md)
- [X] I made sure the code compiles on my machine
- [X] I made sure there are no unnecessary changes in the code
- [X] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [X] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [X] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made (TODO at end of summer)

Marr11317 commented on libmscore/mmrest.cpp:
```diff
@@ -0,0 +1,275 @@
+
+void MMRest::layout()
+{
+    m_number = measure()->mmRestCount();
```
shouldn't this `m_number` be set in the constructor? It seems illogical to me for it to be handled by `layout()`.
Getters and setters?

IsaacWeiss replied:
No, this number needs to be able to change on each relayout, and the object is not destroyed and newly created when the length gets larger or smaller.

Marr11317 commented on libmscore/mmrest.cpp:
```diff
@@ -0,0 +1,275 @@
+    // set clickable area
+    if (score()->styleB(Sid::oldStyleMultiMeasureRests)) {
+        setbbox(QRectF((m_width - m_symsWidth) * .5, -1.0, m_symsWidth, 2.0));
+    } else { // H-bar}
```
forgotten } after H-bar?

IsaacWeiss replied:
Yep

Marr11317 commented on libmscore/mmrest.cpp:
```diff
@@ -0,0 +1,275 @@
+//   resetProperty
+//---------------------------------------------------------
+
+void MMRest::resetProperty(Pid propertyId)
```
wasn't there already a parent class with similar code?

IsaacWeiss replied:
Yes, you're right.

Marr11317 commented on libmscore/style.h:
```diff
@@ -389,6 +388,13 @@ enum class Sid {
     minEmptyMeasures,
     minMMRestWidth,
     mmRestNumberPos,
+    mmRestNumberMaskHBar,
```
Most of these should also be properties I believe.

IsaacWeiss replied:
I don't think so, see [above](https://github.com/musescore/MuseScore/pull/6211#discussion_r439520555).

Marr11317 commented on mscore/editstyle.cpp:
```diff
@@ -716,7 +722,8 @@ const std::map<ElementType, EditStylePage> EditStyle::PAGES = {
     { ElementType::MEASURE,             &EditStyle::PageMeasure },
-    { ElementType::REST,                &EditStyle::PageMeasure },
+    { ElementType::REST,                &EditStyle::PageRests },
+    { ElementType::MMREST,              &EditStyle::PageRests },
```
So there is no page for mmrests after all?

IsaacWeiss replied:
There is! This is it.

mattmcclinch commented:
Your [Git username](https://help.github.com/en/github/using-git/setting-your-username-in-git) is currently set to "Unknown". Please fix this.
Also, you have made a bunch of unneeded changes to `editstyle.ui`. Using [this](https://github.com/mattmcclinch/MuseScore/commit/1e472d9ddcc7aa9f7af21ea83251e79d0560a947#diff-b59d46515d2677f362a2ad7ebb2aa9f4) for the first commit and [this](https://github.com/mattmcclinch/MuseScore/commit/05bca7f6f185ad36bcc6791340f280351f8fa2a3#diff-b59d46515d2677f362a2ad7ebb2aa9f4) for the second commit results in a much nicer diff and a lower likelihood of future conflict.

IsaacWeiss commented on libmscore/mmrest.cpp:
```diff
@@ -0,0 +1,275 @@
+            m_symsWidth += symBbox(sym).width();
+        }
+        if (counter == 1) {
+            m_isOddNumber = true;
```
You're right that it's not useful in terms of program function, but to me, setting the variable here and checking it there makes it more clear that this part of `layout()` is related to that part of `draw()`.

IsaacWeiss commented on libmscore/mmrest.cpp:
```diff
@@ -0,0 +1,275 @@
+
+    // set clickable area
+    if (score()->styleB(Sid::oldStyleMultiMeasureRests)) {
+        setbbox(QRectF((m_width - m_symsWidth) * .5, -1.0, m_symsWidth, 2.0));
```
The fourth parameter of [`QRectF()`](https://doc.qt.io/qt-5/qrectf.html#QRectF-3) is rectangle height. The rest symbols (except for 2 and 3 measures) are always 2sp tall.

Marr11317 replied:
I guess there must be a variable about it somewhere... I don't know where though. Maybe @MarcSabatella the engraving guru knows ;)

IsaacWeiss replied:
Hmm, maybe call the `bbox().height()` for the individual sym and pass it along here?

IsaacWeiss commented on libmscore/mmrest.cpp:
```diff
@@ -0,0 +1,275 @@
+
+void MMRest::write(XmlWriter& xml) const
+{
+    xml.stag("Rest");
```
Sure

IsaacWeiss commented on libmscore/mmrest.cpp:
```diff
@@ -0,0 +1,275 @@
+namespace Ms {
+
+static const ElementStyle mmRestStyle {
+    { Sid::mmRestNumberPos, Pid::MMREST_NUMBER_POS },
```
I feel differently. For the number height, users might want to adjust some particular mmrests for reasons of space, but I don't think there should ever be a score with some H-bars thicker than others. Similarly, old-style should only be applied based on score-wide rules.

MarcSabatella replied:
Still, it doesn't hurt.  To me switching between old style and h-bar on the fly seems a more likely use case, although I'm entirely unfamiliar with the old style and have no insight into how they might normally be used.

MarcSabatella replied:
Given the Inspector still seems to be a bit of a WIP, I wouldn't worry about this right now, but I do still think it's probably a good thing to do eventually.

IsaacWeiss commented on libmscore/mmrest.cpp:
```diff
@@ -0,0 +1,275 @@
+    { Sid::mmRestNumberPos, Pid::MMREST_NUMBER_POS },
+};
+
+const qreal SPACING = 1.5;
```
Sure

MarcSabatella replied:
Yes, definitely a style, probably a property for Inspector.

IsaacWeiss replied:
I'm probably going to do this a different way, in order to get this value into staff spaces. Right now it's actually a multiplier of the width of the previous symbol.

IsaacWeiss commented:
> Your [Git username](https://help.github.com/en/github/using-git/setting-your-username-in-git) is currently set to "Unknown". Please fix this.

I'm not familiar with this. Where do you see "Unknown"? I checked with `$ git config --global user.name` according to the link, and received back `IsaacWeiss` as I expected.

> Also, you have made a bunch of unneeded changes to `editstyle.ui`. Using [this](https://github.com/mattmcclinch/MuseScore/commit/1e472d9ddcc7aa9f7af21ea83251e79d0560a947#diff-b59d46515d2677f362a2ad7ebb2aa9f4) for the first commit and [this](https://github.com/mattmcclinch/MuseScore/commit/05bca7f6f185ad36bcc6791340f280351f8fa2a3#diff-b59d46515d2677f362a2ad7ebb2aa9f4) for the second commit results in a much nicer diff and a lower likelihood of future conflict.

Thanks, I'll check those out.

mattmcclinch commented:
```
$ git log IsaacWeiss/gsoc-2020-old-style-mmrests
commit b832fb7ac3e6a8dcd00525fff7ba7d5bb465160c (IsaacWeiss/gsoc-2020-old-style-
mmrests)
Author: Unknown <isaacaweiss@gmail.com>
Date:   Thu Jun 11 23:58:41 2020 -0400

    Fix #44846: Add old-style mmrests

commit d318db14588ac73fcea9fb70f7c62a8263acd8d8 (IsaacWeiss/gsoc-2020-h-bar-mmre
sts)
Author: Unknown <isaacaweiss@gmail.com>
Date:   Thu Jun 11 18:59:35 2020 -0400

    GSoC 2020: Add style options for H-bar mmrests
...
```

MarcSabatella commented:
It's probably somewhat inevitable that some PR's will be bigger than they otherwise might need to be because of code style changes (renaming variables, etc).  It would help for review purposes to have comments somewhere - probably not in the code - pointing out which changes are intended to be more cosmetic.

IsaacWeiss commented:
Sure, I'll put that comment right here. ;-)

The cosmetic changes involve renaming member variables to start with `m_`, and reordering the header files to put public methods first. That's pretty much it.

MarcSabatella commented:
I get that, but it's kind of a pain to have to read through all the changes in a file only to discover they were *all* just that.  Or, worse, miss the one line of actual content.  So I mean, any large blocks of code that only look like they involve changes but don't, it would be nice to somehow know that up front.

IsaacWeiss commented:
Commits now use @mattmcclinch's versions of editstyle.ui.

mattmcclinch commented on inspectors/view/ui/inspector.cpp:
```diff
@@ -214,6 +214,9 @@ void Inspector::update(Score* s)
                         case ElementType::REST:
                               ie = new InspectorRest(this);
                               break;
+                        case ElementType::MMREST:
+                              ie = new InspectorMMRest(this);
+                              break;
```
This code is not used; `InspectorMMRest` does not exist, and this file has been deleted in https://github.com/musescore/MuseScore/commit/7a05ed77ba29562c132768958621c7dd633f1660#diff-e8bd3356faffc8c421f82c840e014162.

IsaacWeiss commented:
Okay, I think this is done, except for strategic squashing. Any additional review?

IsaacWeiss commented:
Strategic squashing done. I've also fixed the name attached to the first commit. I learned enough about Git today that I can also split that first commit into multiple ones, if desired.

MarcSabatella commented on framework/ui/iconcodes.h:
```diff
@@ -187,6 +187,7 @@ class IconCode
         CHORD_SYMBOL = 0xF352,
         GUITAR_BEND = 0xF353,
         GUITAR_TREMOLO_BAR = 0xF355,
+        MULTIMEASURE_REST = 0xF355,
```
Somehow I doubt it's good to use the same value as tremolo bar :-).  I'll find out soon enough if there is an actual icon added I guess...

mattmcclinch replied:
0xF355 is the code for multimeasure rest in the icon font. As far as I know, there is no glyph yet for guitar tremolo bar.

MarcSabatella replied:
I'm confused, because I *don't* see what the comment in the header says I should see - a change to MuseScoreIcon.ttf.  And yet I see see a multimeasure rest icon in the Inspector.  Maybe it is for something else.  But, guitar tremolo bar shows that same icon in the Inspector.  So definitely something is not right.

MarcSabatella replied:
Sorry, posts crossed.  So the mmrest icon was already there, with code F355, and this file was just lying when it says that was guitar tremolo?

mattmcclinch replied:
There is already a glyph for mulimeasure rest in the icon font. It is at codepoint 0xF355. And it is not for something else; Tantacrul added it specifically so it could be used as the icon for multimeasure rest in the inspector.

MarcSabatella replied:
OK, thanks.  So I guess someone gets to add the tremolo one and change the number in this file, but not the job of this PR :-)

MarcSabatella commented on inspectors/models/notation/mmrests/mmrestsettingsmodel.cpp:
```diff
@@ -0,0 +1,42 @@
+#include "mmrestsettingsmodel.h"
+
+MMRestSettingsModel::MMRestSettingsModel(QObject* parent, IElementRepositoryService* repository)
```
I should mention I am completely unfamiliar with this newfangled model stuff, so I'm learning as I go.  I hope someone more familiar can comment on anything relating to that or to QML specifically.

MarcSabatella commented on inspectors/types/iconnames.h:
```diff
@@ -151,6 +151,7 @@ class IconNameTypes
         CHORD_SYMBOL = 0xF352,
         GUITAR_BEND = 0xF353,
         GUITAR_TREMOLO_BAR = 0xF355,
+        MULTIMEASURE_REST = 0xF355,
```
See previous comment about the number.

MarcSabatella commented on libmscore/layout.cpp:
```diff
@@ -1856,108 +1857,117 @@ System* Score::getNextSystem(LayoutContext& lc)
+            if (m->mmRest()) {
+                undo(new ChangeMMRest(m, 0));
+            }
+            if (m == lastMeasure) {
```
This is type of thing I mean when I say I could use a comment to save the trouble of checking whether you actually changed anything here or just renamed variables.

IsaacWeiss replied:
Ah, yes. Tbh, I should rewind the commit history and make that a separate commit again.

IsaacWeiss replied:
History has been rewritten.

MarcSabatella commented on libmscore/layout.cpp:
```diff
@@ -1976,162 +1986,160 @@ void Score::createMMRest(Measure* m, Measure* lm, const Fraction& len)
         }
-    } else if (ns) {
+    } else if (mmrSeg) {
         // TODO: remove elements from ns?
```
One day something will go wrong where we finally need to understand what is going on here...

MarcSabatella commented on libmscore/layout.cpp:
```diff
@@ -1976,162 +1986,160 @@ void Score::createMMRest(Measure* m, Measure* lm, const Fraction& len)
         }
-    } else if (ns) {
-        // TODO: remove elements from ns?
-        undo(new RemoveElement(ns));
```
...and here

MarcSabatella commented on libmscore/mmrest.h:
```diff
@@ -0,0 +1,61 @@
+    qreal m_numberPos;              // vertical position of number relative to staff
+    std::vector<SymId> m_restSyms;  // stores symbols when using old-style rests
+    bool m_isOddNumber;             // determines whether to add whole rest as final symbol
+    qreal m_symsWidth;              // width of symbols with spacing when using old-style rests
```
To the extent I am not comfortable with any of these, it is that some seems to be just glorified "local" variables generated and cached during layout and not semantically part of the object.  Especially the m_isOddNumber.  Not sure if there is a coding style guideline about this, maybe there should be.

IsaacWeiss replied:
The challenge here is that `layout()` and `draw()` are separate stages, and `draw()` is `const`. Not sure how else to connect them.

IsaacWeiss replied:
No longer using `m_isOddNumber`

MarcSabatella commented on libmscore/edit.cpp:
```diff
@@ -1810,7 +1811,7 @@ void Score::deleteItem(Element* el)
     }
     // fall through
-
+    case ElementType::MMREST:
```
Interesting thought - can there ever be an mmrest on voice > 0?  Would that ever made sense for any reason?

MarcSabatella commented on libmscore/style.cpp:
```diff
@@ -402,6 +401,14 @@ static const StyleType styleTypes[] {
+    { Sid::mmRestHBarVStrokeHeight, "mmRestHBarVStrokeHeight", Spatium(2.0) },
+    { Sid::oldStyleMultiMeasureRests, "oldStyleMultiMeasureRests", QVariant(false) },
+    { Sid::mmRestOldStyleMaxMeasures, "mmRestOldStyleMaxMeasures", QVariant(9) },
+    { Sid::mmRestOldStyleSpacing,   "mmRestOldStyleSpacing",   Spatium(0.85) },
```
Where did this value come from?  Looks wider than what Gould shows, and kind of arbitrary.

IsaacWeiss replied:
Definitely arbitrary. It's the balance I came up with that seems to work well for all the music fonts. Gonville in particular has shockingly thick symbols that don't look good too close together. On the other hand, what do you think of making changing the font automatically change this setting, as well (which would give me an excuse to implement that TODO we talked about before)?

MarcSabatella replied:
Yes, per-font settings would be good.  I've never looked at how this is implemented for the other settings, but I take it is is based on some metadata the accompanies the font and is defined by SMuFL?  And the issue here is that there particular settings are not among those?  Since we have at least some time before MuseScore, I'd suggest actually making a proposal to the SMuFL folks to add what we need.

IsaacWeiss replied:
Opened https://github.com/w3c/smufl/issues/125, but it looks like they've been inactive for some time.

MarcSabatella replied:
Well, I see a post from Daniel from February, and of course it's been a crazy time since then.  We'll see.  They've definitely made other changes at our request before though.

IsaacWeiss replied:
On second thought, I think I'll leave this as is. It's wider than in Sibelius, but narrower than in Finale, and about the same as in Dorico and LilyPond.

MarcSabatella commented on mscore/editstyle.ui:
```diff
@@ -4331,6 +4164,366 @@
+               </widget>
+              </item>
+              <item row="1" column="2">
+               <widget class="QToolButton" name="resetMMRestOldStyleSpacing">
```
Maybe a different step value than 0.01 (which is apparently the default) makes sense here.  Most settings with defaults in this range (somewhere close to 1 sp) have a step of more like 0.1

MarcSabatella commented:
I have completed my review.  Some comments are just observations or questions, a couple suggest actual changes.  Great job, @IsaacWeiss !

IsaacWeiss commented:
> I guess the vtest failures are all actually about the vertical stroke of the h-bar style rests and the scaling change that was made?

Yes, that's all.

MarcSabatella commented:
By scaling, I meant, the scaling *of* the vertical stroke :-)

mattmcclinch commented:
I imagine that https://github.com/musescore/MuseScore/pull/6211/commits/529350451de2fc0f18bd96e108f7626a05c3784d turned out to be not nearly as useful as you were hoping, since the whole rest needs to be drawn at a different y position than the other symbols. I have reworked the code for constructing the symbol list and drawing the symbols, allowing you to drop https://github.com/musescore/MuseScore/pull/6211/commits/529350451de2fc0f18bd96e108f7626a05c3784d altogether. I think it is an improvement. See https://github.com/mattmcclinch/MuseScore/commit/cebef61526ccd824d3b77390e1bd6a33b449ff77.

IsaacWeiss commented:
> I imagine that [5293504](https://github.com/musescore/MuseScore/commit/529350451de2fc0f18bd96e108f7626a05c3784d) turned out to be not nearly as useful as you were hoping, since the whole rest needs to be drawn at a different y position than the other symbols. I have reworked the code for constructing the symbol list and drawing the symbols, allowing you to drop [5293504](https://github.com/musescore/MuseScore/commit/529350451de2fc0f18bd96e108f7626a05c3784d) altogether. I think it is an improvement. See [mattmcclinch@3f9b7f2](https://github.com/mattmcclinch/MuseScore/commit/3f9b7f2dc99f80ef3b142bf44faa47aad55a5853).

You're right, pity. Your code looks almost exactly like what I was originally envisioning, until I discovered `drawSymbols()` and didn't want to duplicate functionality by writing my own loop through the vector. But it does turn out to be much simpler. Thanks.

IsaacWeiss commented:
I believe all review comments have been addressed. The expansion of `drawSymbols()` has been dropped, and @mattmcclinch's suggestions integrated into https://github.com/musescore/MuseScore/pull/6211/commits/8c961524f04ffa6f0fb311f4861617944b404901.

IsaacWeiss commented:
Rebased on latest master

IsaacWeiss commented:
Rebased again

vpereverzev commented on inspectors/view/qml/notation/mmrests/MMRestPopup.qml:
```diff
@@ -0,0 +1,83 @@
+                        isIndeterminate: root.model ? root.model.numberPosition.isUndefined : false
+                        currentValue: root.model ? root.model.numberPosition.value : 0
+
+                        measureUnitsSymbol: "sp"
```
As far as I remember, @Tantacrul was completely against using 'sp' in the inspector, because it makes no sense to 90% of users.

IsaacWeiss replied:
I'll remove it, then.

vpereverzev commented on framework/ui/view/iconcodes.h:
```diff
@@ -187,6 +187,7 @@ class IconCode
         CHORD_SYMBOL = 0xF352,
         GUITAR_BEND = 0xF353,
         GUITAR_TREMOLO_BAR = 0xF355,
+        MULTIMEASURE_REST = 0xF355,
```
2 icons with the same address?

mattmcclinch replied:
0xF355 is the icon code for multimeasure rest. I don’t know where you got the idea that it was the icon code for guitar tremolo bar.

IsaacWeiss replied:
I checked in `MusescoreIcon.ttf` and this is certainly the location of the multimeasure rest icon. There doesn't seem to be a tremolo bar icon at all. For this reason, I'll remove the line `GUITAR_TREMOLO_BAR = 0xF355,` and it can be put back if/when an icon is added.

Tantacrul replied:
@IsaacWeiss @vpereverzev

Can confirm, I've not yet added a guitar tremolo icon.

Tantacrul replied:
Just updated the icon set:
https://www.dropbox.com/s/ip59ren10u69hr7/MusescoreIcon.ttf?dl=0

Guitar Tremolo is now there: 0xF35C

IsaacWeiss replied:
@Tantacrul all taken care of now.

vpereverzev commented on inspectors/view/qml/notation/mmrests/MMRestPopup.qml:
```diff
@@ -0,0 +1,83 @@
+                    }
+                }
+
+                Column {
```
It's better to use InspectorPropertyControl, so users will be able to reset your property value to defaults

IsaacWeiss replied:
Will do.

IsaacWeiss commented:
@vpereverzev anything else?

IsaacWeiss commented:
@vpereverzev @igorkorsukov is there anything else you'd like to see in this PR? I'd like to see this merged and then I'm ready to open a new WIP.
